src/test/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/dynamicPodTemplateStepSupportsRestart.groovy

```diff
@@ -0,0 +1,57 @@
+def call(body) {
+  def config = [:]
+  body.resolveStrategy = Closure.DELEGATE_FIRST
+  body.delegate = config
+  milestone()
+  def yamlPod = '''
+apiVersion: v1
+kind: Pod
+metadata:
+  labels:
+    jenkins: agent
+spec:
+  containers:
+  - name: maven-image
+    command:
+    - cat
+    image: busybox
+    imagePullPolicy: IfNotPresent
+    resources:
+      limits:
+        cpu: 100m
+        memory: 8Mi
+      requests:
+        cpu: 100m
+        memory: 8Mi
+    tty: true
+  dnsPolicy: ClusterFirst
+  restartPolicy: Never
+  securityContext: {}
+  terminationGracePeriodSeconds: 30
+'''
+  def yamlPodHash = '#{project.artifactId}-#{project.version}-maven-image'
+
+  // one build per job to ensure tag consistency
+  try {
+    lock(resource: env.JOB_NAME.split('/')[1],  inversePrecedence: true) {
+      milestone()
+      podTemplate(label: yamlPodHash, instanceCap: 6,  idleMinutes: 60, cloud: 'dynamicPodTemplateStepSupportsRestart', yaml: yamlPod) {
+        node(yamlPodHash) {
```

Comment:
You can just remove the deprecated `label` parameter
```suggestion
  // one build per job to ensure tag consistency
  try {
    lock(resource: env.JOB_NAME.split('/')[1],  inversePrecedence: true) {
      milestone()
      podTemplate(instanceCap: 6,  idleMinutes: 60, cloud: 'dynamicPodTemplateStepSupportsRestart', yaml: yamlPod) {
        node(POD_LABEL) {
```

Reply:
Hi Jesse, thanks for taking a look! I don't understand how that can function without it in a dynamic way (not hard coded into the org setup). 

The class used to work by looking up the pod template and thus available pods by the label. Then it changed to looking up by ID but with enforced randomness in the ID - removing the label would be fine if we could control the ID, but we currently can't control that.

The components that are equivalent to the above snippet are reusable libraries of pod-definition (podTemplate, if you will) and are defined in a totally separate and independent package to the organisation config/setup - this is rewritten completely every boot and is always the same and always works because always the same. On top of the pod template reusable blocks ride a bunch of job types with a many to one relationship to a pod spec, ie, three different  build styles might use one pod spec, and another 3 additional build styles might use a second different pod spec. Different sets of build styles and pod specs are available in different packages with the final jenkins build including those that are useful to a particular team/project.

Without putting these pod specs in the organisation where IMO they do not belong how can you get reusable pods across N jobs that share M build styles that share 1 pod spec/template/definition that is implemented by P agent pods? 

The real version of this has the body passed in from the out side and run as `body()` in place of the println on line 42 of that groovy. 

How does POD_LABEL provide the same experience for us that we had before this got broken and what is the perceived benefit of these items being IDed randomly at creation without it having any bearing on the yaml or name included? Help! :-D

Reply:
> How does POD_LABEL provide the same experience for us that we had before

I am afraid I am unable to follow what you are trying to accomplish or more to the point what you think does not already work. Simple

```groovy
podTemplate(yaml: whatever) {
  node(POD_LABEL) {
    sh 'true' // whatever body
  }
}
```

should work reliably, including across controller restarts, regardless of the number of jobs, concurrent builds, different YAML contents, etc. etc. You should not need to pay any attention to “id”, “label”, etc., or have any global configuration beyond the physical connection to the K8s cluster & namespace (not even that, if the controller itself is running inside K8s and using a service account with appropriate permissions). The above snippet will create a `Pod` with the definition you specify, wait for its agent to connect to the controller, do work, then delete the pod and clean up, with no interference with other builds or jobs or global settings.

In the future we could introduce a separate step that bypasses the Jenkins queue altogether and makes things even simpler:

```groovy
pod(yaml: whatever) {
  sh 'true'
}
```

by analogy with https://www.jenkins.io/doc/pipeline/steps/docker-slaves/#dockernode-allocate-a-docker-node but for now the `POD_LABEL` idiom is the glue needed to make dynamic pod configuration work with the stock `node` step.

Reply:
Hey Jesse, we're somewhat in the same boat here :-) You don't understand what we're trying to do and we don't understand why you're not letting us do it :-D But the above clarifies your angle so hopefully I can build on that and explain what our intent is:

> The above snippet will create a Pod with the definition you specify, wait for its agent to connect to the controller, do work, then delete the pod and clean up, with no interference with other builds or jobs or global settings.

This is 100% *not* what we want :-D The idea is that we specify the definition in such a way that it matches existing pods and can then just use them. The result of the change Vincent made was that it doesn't match or not match, it just fails.

It seems like you've interpreted that groovy file as having the intent to create a bespoke single use pod, use it, and dump it - is that right? That's not the intent, and this setup worked flawlessly for years before the commit that changed from:

```
set label in unique to def way
lookup by label in KubeSlave class
just work
```

to

```
set label in unique to def way, get random id
lookup by id in KubeSlave class
fail
```

Allowing us to set that ID would restore perfect clean functionality for us - but both of you seem pretty heavily against that, so I need to understand what a solution that meets your idea of how it **should** be used AND meets our idea of an acceptable way to use it is, and if that can be found, rework our stuff and roll it out everywhere. And if not, maybe just fork it, add the one line, and hope no further breakage occurs. Though I really really really don't want to do that for a lot of reasons.

One of the reasons we don't want brand new pods every time is the cache that is held in the pod between builds will be empty every time and every build will be ultra slow, not just those first thing in the working day. Even if/when we do something to improve the cache seeding situation that seeding will be a fairly heavy overhead (largeish data copy) for a new pod and not something we'd want to do hundreds or thousands of times per day. 

Our goal is:

1. Our jenkins(actually many of them) is fresh/new every time it boots, always the same, eg  - those random IDs appear to break that - but perhaps there's another way
2. Our builds specify the pod style they use in a self contained way 
3. Pods are reused throughout the day until their inactivity timeout is reached and they're killed and the cluster can down scale - hot caches, fast builds - set to 1 hour inactivity so if used regularly through the day, life span of approx 9 or 10 hours for a single-time-zone team
4. The jenkins assembly can be packaged together as a composition including sets of rules/pod-defs that can immediately be used
5. After a restart IF the pod defs are the same, the old pods should be successfully reused
6. Ideally minimal rework of our stack as we're pretty happy with how it is at the moment - however we're open to change, it's just not clear what the new acceptable solution would look like yet.

I don't understand how any fresh-on-boot system can be consistent and can match the pod template when it runs if it's not possible to create those and specify the ID or have the ID derived from a hash of the contents? That would be fine, too.

Maybe it IS possible but in a different way and not in the build command spec like now where it goes via the PodTemplateStep object (created confusingly by podTemplate() call) and not directly into a PodTemplate object where that ID can already be set.

I hope that clarified a few things for you as your post did for me, and I look forward to your response! :-D

Reply:
> we don't want brand new pods every time

The `kubernetes` plugin is designed to create a pod for every build (or more precisely, every `node` block), and then throw it away once the build (block) exits. Reusing pods is not a supported use case. (There is an option to keep a pod after exiting, but this is solely for diagnostic purposes, and even that is a dubious practice.)

> the cache that is held in the pod between builds will be empty every time

You can specify a persistent volume to reuse between pods, assuming the storage class supports concurrent mounts. This is generally a quite fragile setup, though (special requirements on volume drivers, UID mismatches, file locking issues, ad nauseam) so you are generally better off using a mirror (if dependency downloads are the problem) or a distributed-friendly build system with a cache server (if incremental builds are what you are looking for).

Reply:
BTW this is a good example of the importance of starting off explaining what problem you are trying to solve, not only what solution you attempted.

Reply:
Hey Jesse!

From your first paragraph, can I infer that this was never the goal of the kubernetes plugin since the start? And that therefore it only ever worked perfectly and efficiently in a clean and repeatable way by accident or chance? That's super unfortunate. Who were the group of people who founded this plugin project?

Jenkins has had reusable agents that can run on EC2 hosts from AMIs for years and years, our use case is **exactly** the same as that, but pods instead of EC2s because of the massive benefits of doing this in Kube including cloud agnostic use. What is the rational behind explicitly excluding this use case in the kube plugin and marking this use case as unsupported? 

I'm pretty solid on how kube works, both from a consumption point of view, and an under the hood point of view. I've been using Jenkins since it before it was called that in 08. I'm currently handling more than a dozen clusters across two clients and ourselves and dozens of environments running inside those clusters. Cattle is not new to me, any EC2s I've managed in the past and had any control over were also cattle, kill and replace, and enjoy the steak. Our current Kubernetes Jenkins agents are cattle, too. They are born in the morning and die in the evening and tomorrows ones are all new. 

In fact, the kubernetes plugin correctly supports this scenario to this minute provided that the master/controller is not restarted. That is the only issue, it works cleanly and has done for years, with that one caveat introduced by the aforementioned commit. 

As you state, using persistent volumes is not a reasonable solution and anything that does support concurrent mounts could result in repo cache corruption, which is a bad thing. Most of our stacks have a local or almost local mirror already in place and this does help, but it's about an order of magnitude slower than a local cache. I'm 100% sure you're familiar with maven but given I've been using m2/m3 for 14 years now across multiple countries and businesses I'll just reiterate what occurs in each case for anyone else stumbling upon this thread of conversation:

local cache populated:

- ranges are resolved from the remote as needed for a particular build
- if the range specification match is present on the host it is consumed immediately
- if not, the set of particular artifacts are downloaded as needed on a per plugin per plugin execution basis
- although some parallelism is afforded with specific configurations, this is largely a sequential operation
- as such connection latencies and handshakes kill time and other things, are, by the nature of maven, inherently pulled down as needed, not all up front. 
- this is normal for internal artifacts that are probably referred to by ranges, definitely available on the in-house remote, and fairly quick to get, because inherently there's few generated by human actions per hour/day, no matter how fast the coders are and the PRs get reviewed thoroughly and tests run and merges occur
- with proper cascades in place each build is typically only pulling down the thing that got built immediately before assuming it's not on the same node

new pod, no local cache, but fastish cluster cache:

- ranges are resolved from the remote as needed for a particular build
- ranges never get a cache hit and every plugin, library, and other artifact has to be downloaded EVERY time
- this happens mostly sequentially as each build phase begins and throughout the build almost to the end
- although it's faster than pulling from central because all central artifacts required are cached in the cluster, due to the latencies and staged nature of it, it's still much slower
- on top of that, many of our builds currently finish faster than a kube agent pod can come up and register with Jenkins and begin its work - in these cases the performance would be worse than half even if the cache were full and more than an order of magnitude with the cache empty and a local cluster cache feeding the build

I'm curious how you can justify calling reusing pods or even keeping them alive to inspect build directory files that vanish when they go away, a dubious practice. To me, this is no different than reusing web server pods to fulfil REST or SOAP  requests - when there's more load, there's more of them, when there's less load, some go away, they're cattle, but you keep them around while you need them and having them live for a day or a week is typically not an issue or a benefit, it's just what it is. In contrast anything with a cache benefits from being kept around longer, no matter that cache type. 

I think the first sentence covered the problem I was trying to solve AND provided an almost-certainly-working solution for it. The fact of the matter seems to be that you were both blinkered by how used to throwing away pods **every** time you both are and didn't even consider/see the standard keep them as long as you need them paradigm.

I would strongly urge you to reconsider pod reuse being unsupported as it's already supported with the exception of restarts (which we can easily fix) and is a technically and philosophically sound practice on every level I can think of. 

Can you give any kind of justification for why the EC2 agent pool like model is not appropriate in kube? I'd love to hear it.

And can you specify what you mean by `a distributed-friendly build system with a cache server`, please? I'm curious.

Reply:
> it only ever worked perfectly and efficiently in a clean and repeatable way by accident or chance

As far as I am aware. (Evidently was no test coverage for such a usage pattern, for example.)

> reusable agents that can run on EC2 hosts from AMIs

For VMs, sure, because these have a boot process. K8s uses containers which are lightweight and the intention is to use one per build(-ish).

> if the range specification match

Off-topic but Maven version ranges are a really bad idea IMHO.

> faster than a kube agent pod can come up and register with Jenkins and begin its work

Interesting, I typically see ~10s times for this. Of course some Maven builds of tiny projects are indeed faster than that.

> this is no different than reusing web server pods

Well, depends on your setup. “Web servers” are typically stateless. Reusing an environment to run CI builds is a much riskier proposition _in general_. (Especially with Maven given its historical issues: lack of thread safety on `~/.m2/repository/`, `mvn install` on non-`SNAPSHOT` POMs quietly clobbering artifacts which would normally have been downloaded from a repo, etc.) Builds might not be fully reproducible from sources, depending on how careful you are. There are also security risks, if you do not have homogeneous permissions.

> can you specify what you mean by _a distributed-friendly build system with a cache server_

https://bazel.build/ for one. https://docs.cloudbees.com/docs/cloudbees-build-acceleration/latest/ (my employer but not my area of expertise). https://docs.gradle.org/current/userguide/build_cache_use_cases.html#share_results_between_ci_builds or even https://docs.gradle.com/enterprise/maven-build-cache/ supposedly.

Reply:
> As far as I am aware. (Evidently was no test coverage for such a usage pattern, for example.)

How about I add a decent suite that covers this standard and very much still valid and appropriate and not at all dubious practice so that we can then find a way to fix it to your and Vincent's liking and keep it from breaking in future?
 
> For VMs, sure, because these have a boot process. K8s uses containers which are lightweight and the intention is to use one per build(-ish).

Not having a boot process is not a good reason to discard after every single use every single time. Many containers do have an init process, and Jenkins agent containers are no exception from this having to register with the mothership before they can do anything at all.

> > if the range specification match
> 
> Off-topic but Maven version ranges are a really bad idea IMHO.

Indeed, off topic, however ranges **ARE** a **REALLY** bad idea when applied to artifacts that you do not control internally. For artifacts that are controlled internally, they provide a perfect balance between fluidity and control and with the addition of just a little discipline bumping versions when making breaking changes so as to exclude new different builds from existing consumption ranges, they provide an unmatched high productivity development experience. 

> Interesting, I typically see ~10s times for this. Of course some Maven builds of tiny projects are indeed faster than that.

Yep, and in any good project that's more or less all you have. Small projects and composition. Legacy monoliths can take an hour in the worst case, and regularly half that. I hope they're all gone by the time my funeral service is held! :-D

> Well, depends on your setup. “Web servers” are typically stateless.

True, but alas, it's not much different, there's just a cache. If there was no cache it'd be exactly the same. But there is a cache. And many apps have caches, too. 

> Reusing an environment to run CI builds is a much riskier proposition _in general_. (Especially with Maven given its historical issues: lack of thread safety on `~/.m2/repository/`, `mvn install` on non-`SNAPSHOT` POMs quietly clobbering artifacts which would normally have been downloaded from a repo, etc.) Builds might not be fully reproducible from sources, depending on how careful you are. 

Well, I'm not sure how you're using Maven, but most of that is a total non issue for us. To reproduce a build full of ranges from source you need a set of range resolutions, those resolutions are baked into our deployed poms such that it's trivial to generate a pom to reproduce a build if needed, however it's rarely needed, always move forward, always develop forward. If you are close to production with your source, fixing it and producing a new build that passes all tests and rolling it straight out is a preferable option almost every time.

> https://bazel.build/ for one. https://docs.cloudbees.com/docs/cloudbees-build-acceleration/latest/ (my employer but not my area of expertise). https://docs.gradle.org/current/userguide/build_cache_use_cases.html#share_results_between_ci_builds or even https://docs.gradle.com/enterprise/maven-build-cache/ supposedly.

Thanks. First one seems aimed at monolithic source trees. Second one not sure what it's aimed at, but sounds complicated. Simple is good for us. Last one is the antithesis and a step backward toward ant. There's a lot to be said for having a standard that any J dev can sit down and understand. Ant was never that just as ansible is never the right choice for infra and gradle is the new ant. 

So now that we appear to be 100% on the same page, is the verdict that I can go ahead and work on tests to cover this common scenario so it is available generally, including documentation, to users of this plugin? Or is the verdict a hard no, this plugin only supports transient pods going forward even though it supported reusable pods in full until about 8 months ago and still supports them provided you can keep your master/controller pod up contiguously? I hope the former, if so, I'll get stuck in on Tuesday (or possibly before) making that happen and look forward to harsh/honest feedback in that process :-D

Cheers!

Reply:
> it supported reusable pods in full

But only with a single executor? Whereas the EC2 plugin (say) would let you set up a disposable but possibly long-lived agent with as many executor slots as the hardware makes feasible. More efficient for very short builds, at the cost of a higher risk of cross-build interference and other maintenance overhead. That is in line with how the `Cloud` API was originally conceived in Jenkins. The `kubernetes` plugin still uses this interface, and `Queue`, though it has to hack around many outdated assumptions. We have considered an approach more like [this step](https://github.com/jenkinsci/docker-plugin/blob/87477e9ed176cfdf3132745db65153e0b1d7626b/src/main/java/io/jenkins/docker/pipeline/DockerNodeStepExecution.java#L102-L109), where the container is allocated and released directly in the control flow of a Pipeline step.

> is the verdict that I can go ahead

Up to @Vlatombe—I am not a maintainer. My _personal_ inclination would be to say no, this VM-like use case with reusable agents is outside the design space of this plugin, and given the likely maintenance burden relative to the apparently rare demand it is not worth expanding the scope of what is supported (tested & documented). It is often very difficult with this plugin to draw a clear line between deliberately supported use cases and creative installations that happen to work, because it offers rather low-level facilities (no other K8s CI tool I am aware of lets you freely define `Pod` YAML).

Reply:
My take on this, with the given plugin state:
* Static pod templates defined in the Jenkins UI or through JCasc can be reused (I think this is actually the original use case introduced in the plugin)
* Dynamic pod templates defined in Jenkinsfiles or shared libraries through the DSL are single-use. They're not intended to be reused or shared with other projects or builds.

The fact dynamic pod templates (or corresponding pods) are accessible to other jobs is a undesirable side-effect of using the Cloud API. Unfortunately this is what we have today, and as Jesse mentioned in his earlier response this is something we want to eliminate if we have the opportunity.

For your use case, I think it can fit with the current plugin if you manage static pod templates using the jcasc aproach. I'm not in favour to make any plugin change that would facilicate re-use of pods if they have been defined in a Jenkinsfile.

Reply:
> For your use case, I think it can fit with the current plugin if you manage static pod templates

IIUC no, the request was to reuse _pods_, not just pod templates, for performance reasons. Certainly reusing pod _templates_ across jobs is well supported (even if we would be inclined to recommend the `podTemplate` step going forward), but that is still assuming one pod per `node` block.

Reply:
See https://github.com/jenkinsci/kubernetes-plugin/blob/4068c524f534c30c3c480c91c48426b3d5f2acf8/src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesSlave.java#L538-L544

Reply:
`OnceRetentionStrategy` is the normal usage, but the option to use `CloudRetentionStrategy` suggests that the long-lived-agent style was deliberately added—albeit without tests, documentation, or the ability to use multiple executors? Bisecting says in #91, whose description seems to match this request.

Reply:
Hi Jesse, Vincent, long time no talk! I did see all of these updates but have been busy with other stuff.

This is still burning everyone, however, and the number of people it's burning keeps growing. Please stop calling it a rare demand. I speak for 50+ people across 7 or so organisations when I comment here. I'd rather not repeat that a third or subsequent time. This is a big deal for a lot of people across two banks and various other companies.

> More efficient for very short builds, at the cost of a higher risk of cross-build interference and other maintenance overhead.

Please, this is your personal perception and not any part of our reality. These aspects are managed properly and never give us any trouble.

> this VM-like use case with reusable agents is outside the design space of this plugin

It's not "vm like" it's normal for well composed software that is assembled from small reusable pieces.

You have personally, and I commend you for your curiosity and diligence in researching it, proven that that's not the case. It was added to this plugin by the *founder* of the plugin, and for good reason. Let's call that the design space. There are another dozen people wanting it on the various links you provided, in addition to the multiple dozens I bring to the table.

> I'm not in favour to make any plugin change that would facilicate re-use of pods if they have been defined in a Jenkinsfile.

Our Jenkinsfiles are one or two line, paraphrasing:

```
buildACertainWayWithACertainAmountOfResources
triggerOtherDownstreamBuilds(["X", "Y", "Z"])
```

And to your other comment:

> Dynamic pod templates defined in Jenkinsfiles or shared libraries through the DSL are single-use. They're not intended to be reused or shared with other projects or builds.

How are pod templates defined in shared libraries that are static with respect to a thousand builds not static in reality? 

I understand that you can't guarantee that, but can you make the assumption that someone neck deep in configuring this has the talent to ensure they don't hang themselves with the rope they're given, and just give them the rope?

All of our pod templates are named and versioned, unique and immutable. If they don't match, you get a new pod. If they do, during a master/controller run, or even after a restart before the fateful change, they get reused. 

I'd like to take this moment to remind you that this is ONLY across restarts that there is now an issue. It STILL works as required during a single master node execution cycle. Please do not cripple it any further. 

My change just lets me control the ID such that it's not random and is deliberate and meaningful in a controlled and safe way. For existing users, this makes no difference unless they're currently seeing a "field doesn't exist" of the same name by random chance. Only then could behaviour change. 

My change simply matches the available-to-me API with the underlying object API, it doesn't add anything new, it just exposes a TINY bit more control that wasn't previous required when the default behaviour made sense and wasn't broken. I understand the desire to make them unique for your throw-away style of builds, but please understand our desire to keep them matching and reusable and look-up-able for our builds that work the same exact way, guaranteed by their own semantics, whether there is cache available, or not, just MUCH slower when not. 

Please let us move on and let this PR die in a good way.